Source/WebCore/platform/audio/SharedAudioDestination.cpp

```diff
@@ -0,0 +1,249 @@
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "SharedAudioDestination.h"
+
+#include "AudioUtilities.h"
+#include <wtf/WTFSemaphore.h>
+#include <wtf/WeakPtr.h>
+#include <wtf/WorkQueue.h>
+
+namespace WebCore {
+
+class SharedAudioDestinationAdapter : public RefCounted<SharedAudioDestinationAdapter>, public CanMakeWeakPtr<SharedAudioDestinationAdapter>, public AudioIOCallback {
+public:
+    using AudioDestinationCreationFunction = SharedAudioDestination::AudioDestinationCreationFunction;
+    static Ref<SharedAudioDestinationAdapter> ensureAdapter(unsigned numberOfOutputChannels, float sampleRate, AudioDestinationCreationFunction&& ensureFunction);
+
+    void addRenderer(SharedAudioDestination&, CompletionHandler<void(bool)>&&);
+    void removeRenderer(SharedAudioDestination&, CompletionHandler<void(bool)>&&);
+
+    unsigned framesPerBuffer() const
+    {
+        return m_workBus->length();
+    }
+
+private:
+    using AdapterKey = std::tuple<unsigned, float>;
+    using AdapterMap = HashMap<AdapterKey, WeakPtr<SharedAudioDestinationAdapter>>;
+    static AdapterMap& sharedMap();
+
+    SharedAudioDestinationAdapter(unsigned numberOfOutputChannels, AudioDestinationCreationFunction&&);
+
+    void render(AudioBus* sourceBus, AudioBus* destinationBus, size_t framesToProcess, const AudioIOPosition& outputPosition) final;
+    void isPlayingDidChange() final { }
+
+    void configureRenderThread(CompletionHandler<void(bool)>&&);
+
+    Ref<AudioDestination> m_destination;
+    Ref<AudioBus> m_workBus;
+    Ref<WorkQueue> m_configurationQueue;
+
+    Lock m_renderLock;
+    Semaphore m_configurationSemaphore;
+
+    // Only access on MainThread
+    Vector<RefPtr<SharedAudioDestination>> m_renderers;
+    bool m_started { false };
+
+    // Protected by m_renderLock:
+    bool m_needsConfiguration { true };
+    Vector<RefPtr<SharedAudioDestination>> m_newRenderers;
+    Vector<RefPtr<SharedAudioDestination>> m_configuredRenderers;
+};
+
+auto SharedAudioDestinationAdapter::sharedMap() -> AdapterMap&
+{
+    static NeverDestroyed<AdapterMap> map;
+    return map;
+}
+
+Ref<SharedAudioDestinationAdapter> SharedAudioDestinationAdapter::ensureAdapter(unsigned numberOfOutputChannels, float sampleRate, AudioDestinationCreationFunction&& ensureFunction)
+{
+    auto key = std::make_tuple(numberOfOutputChannels, sampleRate);
+    auto results = sharedMap().find(key);
+    if (results != sharedMap().end() && results->value)
+        return Ref { *results->value };
+
+    Ref newAdapter = adoptRef(*new SharedAudioDestinationAdapter(numberOfOutputChannels,  WTFMove(ensureFunction)));
+    sharedMap().set(key, newAdapter);
```

Comment:
Why `set()` and not `add()`? My suggestion would be to use the more efficient `add()` and add an assertion that the key is not already present in the map.

Reply:
Also, any reason you're not doing this in the constructor like we usually do?

Reply:
I see you adding to the map but never removing, seems fishy.

Reply:
> Why set() and not add()? My suggestion would be to use the more efficient add() and add an assertion that the key is not already present in the map.

Because the key might be in the map, but the value is WeakPtr/null. `add()` would leave the null value in place rather than replace it with a valid value.